### PR TITLE
Grep options can be passed by ENV to both client and server

### PR DESCRIPTION
**Pull request by keyscores**
This proposes a feature for: https://github.com/DispatchMe/meteor-mocha/issues/7

To pass state between terminal, web reporter, server and client perhaps the most meteor-ish way is using a collection and pub/sub. Meteor.settings.public, the current implentation, could not work for a web reporter that is expected to rerun tests with grep patterns (like practicalmeteor:mocha). 

Plus in this commit command line exported ENV are consolidated and added to a one document collection.

From readme:
### Run tests inclusively (grep), and exclusively (invert)

To run any tests which names match a pattern, add the environment variabel `MOCHA_GREP=your_string`. This will apply to both client and server tests.

To exclude any tests, you must use the grep option above plus `MOCHA_INVERT=true` or`MOCHA_INVERT=1`. For example, to exclude tests named 'TODO:' ( which you may want to exclude from your continuous integration workflow) you would pass at runtime `MOCHA_GREP=your_string MOCHA_INVERT=1`

### CODE REVIEW

**Comment by keyscores** [comment]
also fixes https://github.com/DispatchMe/meteor-mocha/issues/5

**Comment by aldeed** [comment]
@keyscores Can you explain why saving the config on settings doesn't work? I don't see anything happening in the collection that couldn't happen the old way, and I'd like to avoid extra dependencies on mongo and such if they aren't needed.

I can see why we might need a server method for the client to pass info back to the server. That could fix the current limitation that changing a client-only file in watch mode doesn't rerun client tests. But I'm not understanding the benefit of the collection/sub part.

**Comment by keyscores** [comment]
It would work, we could use settings and a server method. However, if there is going to be an API for a web reporter a dependency on mongo will be needed anyway (I haven't submitted a pull request for that part yet). So considering this, I preferred using a collection for the settings as well, since we can get reactivity and the callbacks from subscribe.

**Comment by aldeed** [comment]
Fair enough, but I don't see any benefit. It seems like a bunch of extra code vs the old way. I don't think the server will ever restart/rerun tests without the browser also refreshing (or being reloaded entirely if using the headless browsers), so reactivity isn't necessary.

We do have to consider what happens when only the browser page refreshes, but that will rerun all the client code, so no reactivity needed there either.

We can use pub/sub to send server results to the client reporter without using a server side collection. I'd prefer to avoid needing mongo. It should just use the publish function API, added/changed/removed/ready.

I'm still on the fence a little about whether it's a good idea to bring all this complexity into this package. Maybe it could be in dispatch:mocha-browser and then only people who want html reporter would need that. In my experience, not too many people want it.

**Comment by keyscores** [comment]
I get wanting a simple package.  I'm working on this to try to merge the features of practicalmeteor and dispatch. If you think that is a worthwhile goal, then I think using collections doesn't add so much risk.

To be clear I don't think dispatch should have a webreporter, but provide the data for one. And using a non-mongo solution would tie you to providing it.

In terms of implementation, yes typically the client tests would always run in the webreporter, but I'm wondering if in nightmare for example, we can set a window variable to only run in that case.

**Comment by keyscores** [comment]
See now the previous version without using collections. Web reporter would have to set new args and get state by method.

**Comment by keyscores** [comment]
@aldeed Any thoughts on this pull request?

File: client.js [hunk 1]
```diff
@@ -8,7 +8,16 @@ function runTests() {
...
-  mocha.reporter(Meteor.settings.public.CLIENT_TEST_REPORTER || 'spec');
+
+  let runtimeArgs = Meteor.settings.public.runtimeArgs
+  if (runtimeArgs.mochaOptions.grep) { mocha.grep(runtimeArgs.mochaOptions.grep) }
+  if (runtimeArgs.mochaOptions.invert) { mocha.options.invert = runtimeArgs.mochaOptions.invert }
+
+  let clientReporter  = runtimeArgs.mochaOptions.reporter
+  if (runtimeArgs.mochaOptions.clientReporter){
+    clientReporter  = runtimeArgs.mochaOptions.clientReporter
+  }
```
Comment (aldeed) [inline client.js]: I would simplify/rewrite this to:

```js
const { mochaOptions } = Meteor.settings.public.runtimeArgs || {};
const { clientReporter, grep, invert, reporter } = mochaOptions || {};

if (grep) mocha.grep(grep);
if (invert) mocha.options.invert = invert;
mocha.reporter(clientReporter || reporter);
```

File: server.js [hunk 2]
```diff
@@ -1,17 +1,16 @@
...
-});
+let runtimeArgs = setArgs();
 
+const shouldRunClientTests = runtimeArgs.runnerOptions.runClient
+const shouldRunServerTests = runtimeArgs.runnerOptions.runServer
+const shouldRunInParallel = runtimeArgs.runnerOptions.runParallel
+let serverReporter  = runtimeArgs.mochaOptions.reporter
+if (runtimeArgs.mochaOptions.serverReporter){
+  serverReporter  = runtimeArgs.mochaOptions.serverReporter
+}
```
Comment (aldeed) [inline server.js]: You can reduce the code here using destructuring, similar to my suggestion in the client file.

**Pull request closed without merging**